lib/Sema/CSApply.cpp

```diff
@@ -107,6 +107,17 @@ Type Solution::computeSubstitutions(
   auto lookupConformanceFn =
       [&](CanType original, Type replacement, ProtocolType *protoType)
           -> ProtocolConformanceRef {
+    // Replacing type variable with archetype (if possible) is required
+    // here because type variables are never going to produce concrete
+    // conformance nor would have any dependent protocols exposed directly,
+    // which means that later assertion is going to be triggered even in
+    // case where type variable is related to archetype which confirms to
+    // required interface.
+    if (auto typeVar = replacement->getAs<TypeVariableType>()) {
```

Comment:
@DougGregor Ok, so I've (re-)realized that simply using simplifyType there is not going to work because in `matchWitness` `solveSingle` allows free type variables, so in substitutions map type variable basically is going to be replaced with itself... The code I'm talking about is at https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckProtocol.cpp#L1244